### PR TITLE
Add some additional ics2rem functionality.

### PR DESCRIPTION
**Pull request by Xitian9**
Add various functionality to ics2rem, including:
- Add a tail to every MSG, which follows the summary and preceeds the description. This is intended to add things like ‘%b’ to the end of MSGs.
- Add deltas for advance warning on dates, as well as deltas and repeats for times.

I've been doing these things with sed scripts so far, but that's clearly the wrong solution. I think this fix adds the functionality without adding too much complexity.

### CODE REVIEW

**Comment by Xitian9** [comment]
Come to think of it, there may not be a need for the trepeat parameter, since it could just be bundled into the tdelta (they're both strings which are concatenated immediately after one another). It would add a bit to the cognitive load of determining what the parameters mean, but would simplify the code. Do others have opinions on which approach is best?

File: remind.py [hunk 1]
```diff
@@ -347,20 +347,28 @@ def _gen_msg(vevent, label):
...
             rem.append('%%"%s%%"' % ' '.join(msg))
-            rem.append(vevent.description.value.strip())
         else:
             rem.append(' '.join(msg))
+        if tail:
+            rem.append(tail)
+
+        out = ' '.join(rem)
+        if has_desc:
+            out += sep + vevent.description.value.strip()
```
Comment (jspricke) [inline remind.py]: I would propose to append the string to the list and move the join back to it's old place.

File: remind.py [hunk 2]
```diff
@@ -347,20 +347,28 @@ def _gen_msg(vevent, label):
...
+        return out.replace('\n', '%_').replace('[', '["["]')
 
     @staticmethod
     def _abbr_tag(tag):
         """Transform a string so it's acceptable as a remind tag. """
         return tag.replace(" ", "")[:48]
 
-    def to_remind(self, vevent, label=None, priority=None, tags=None):
+    def to_remind(self, vevent, label=None, tail=None, sep=" ", priority=None,
+                  tags=None):
```
Comment (jspricke) [inline remind.py]: Could you keep the old argument order, so that we keep the API stable?

File: remind.py [hunk 3]
```diff
@@ -358,17 +358,22 @@ def _gen_msg(vevent, label, tail, sep):
...
-            out += sep + vevent.description.value.strip()
+            out += sep + Remind._rem_clean(vevent.description.value)
 
-        return out.replace('\n', '%_').replace('[', '["["]')
+        return out
 
     @staticmethod
     def _abbr_tag(tag):
-        """Transform a string so it's acceptable as a remind tag. """
+        """Transform a string so it's acceptable as a remind tag."""
```
Comment (jspricke) [inline remind.py]: Please don't change unrelated lines (or put them in a separate commit)

File: remind.py [hunk 4]
```diff
@@ -358,17 +358,22 @@ def _gen_msg(vevent, label, tail, sep):
...
 
-    def to_remind(self, vevent, label=None, tail=None, sep=" ", priority=None,
-                  tags=None):
+    @staticmethod
+    def _rem_clean(rem):
+        """Transform newlines and '[' in string."""
+        return rem.strip().replace('\n', '%_').replace('[', '["["]')
+
+    def to_remind(self, vevent, label=None, tail=None, sep=" ", delta=None,
+                  tdelta=None, trepeat=None, priority=None, tags=None):
```
Comment (jspricke) [inline remind.py]: This is unrelated to the patch as well.

File: remind.py [hunk 5]
```diff
@@ -524,9 +539,12 @@ def ics2rem():
...
     parser = ArgumentParser(description='Converter from iCalendar to Remind syntax.')
     parser.add_argument('-l', '--label', help='Label for every Remind entry')
-    parser.add_argument('--tail', help='Tail for every Remind entry')
-    parser.add_argument('--sep', default=" ",
+    parser.add_argument('--tail',        help='Tail for every Remind entry')
+    parser.add_argument('--sep', default=' ',
                         help='String to separate summary from description')
+    parser.add_argument('--delta',       help='delta for Remind date')
+    parser.add_argument('--tdelta',      help='delta for Remind time')
+    parser.add_argument('--trepeat',     help='repeat for Remind time')
```
Comment (jspricke) [inline remind.py]: can you add some more help here?

File: remind.py [hunk 6]
```diff
@@ -524,9 +539,12 @@ def ics2rem():
 
     parser = ArgumentParser(description='Converter from iCalendar to Remind syntax.')
     parser.add_argument('-l', '--label', help='Label for every Remind entry')
-    parser.add_argument('--tail', help='Tail for every Remind entry')
-    parser.add_argument('--sep', default=" ",
+    parser.add_argument('--tail',        help='Tail for every Remind entry')
```
Comment (jspricke) [inline remind.py]: Please use only one space, don't vertical align (this just breaks with new commits!). Also, this should be in the first commit ;).

File: remind.py [hunk 7]
```diff
@@ -358,17 +358,22 @@ def _gen_msg(vevent, label, tail, sep):
...
     @staticmethod
     def _abbr_tag(tag):
-        """Transform a string so it's acceptable as a remind tag. """
+        """Transform a string so it's acceptable as a remind tag."""
         return tag.replace(" ", "")[:48]
 
-    def to_remind(self, vevent, label=None, tail=None, sep=" ", priority=None,
-                  tags=None):
+    @staticmethod
+    def _rem_clean(rem):
```
Comment (jspricke) [inline remind.py]: This sounds like a good idea, can you put it in an extra commit?

File: remind.py [hunk 8]
```diff
@@ -372,9 +372,18 @@ def _rem_clean(rem):
         """Transform newlines and '[' in string."""
         return rem.strip().replace('\n', '%_').replace('[', '["["]')
 
-    def to_remind(self, vevent, label=None, tail=None, sep=" ", delta=None,
-                  tdelta=None, trepeat=None, priority=None, tags=None):
-        """Generate a Remind command from the given vevent"""
+    def to_remind(self, vevent, **kwargs):
```
Comment (jspricke) [inline remind.py]: This changes the API quite a bit, also, using strings to access the dict tricks some static analysis tools. I would propose to go back to the old way.

**Comment by Xitian9** [comment]
Will do. Sorry for the bad commit style.

Do you have an opinion on merging the tdelta and trepeat parameters, or keeping them separate? Barring any objections, I'll just merge the two.

File: remind.py (hunk 1 above)
Comment (Xitian9) [inline remind.py]: The reason I've done it this way is that I don't necessarily want a space there. The situation I want is where an entry with summary ‘Foo’ and description ‘Bar’ is transformed not to `%"Foo%" Bar`, but rather to `%"Foo%"%_Bar`.
With the append functionality, this would instead become `%"Foo%" %_ Bar`. Then again, this may not be a big enough problem to justify the slight mess it makes in the code.

**Comment by Xitian9** [comment]
Cleaning this up was messy enough that I've created a new pull request #7. Closing.
